### PR TITLE
Add support for STM32 ADC power management

### PR DESCRIPTION
**Pull request by gautierg-st**
This PR adds support for the power management of the STM32 ADC.
It also adds a sample to test power consumption.
Without PM, the measured consumption is 475µA, with PM, about 25µA

### CODE REVIEW

File: samples/boards/stm32/power_mgmt/adc/prj.conf [hunk 1]
```diff
@@ -0,0 +1,6 @@
+CONFIG_PM=y
+CONFIG_PM_DEVICE=y
+CONFIG_PM_DEVICE_RUNTIME=y
+CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=n
+CONFIG_ADC=y
+#CONFIG_DEBUG=n
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/prj.conf:6]: Either #CONFIG_DEBUG=y or CONFIG_DEBUG=n

File: samples/boards/stm32/power_mgmt/adc/prj.conf [hunk 2]
```diff
@@ -0,0 +1,6 @@
+CONFIG_PM=y
+CONFIG_PM_DEVICE=y
+CONFIG_PM_DEVICE_RUNTIME=y
+CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=n
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/prj.conf:4]: Did you check behavior with `CONFIG_PM_DEVICE_RUNTIME_EXCLUSIVE=y`

File: samples/boards/stm32/power_mgmt/adc/boards/nucleo_wb55rg.overlay [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+	channel@4 {
+		reg = <4>;
+		zephyr,gain = "ADC_GAIN_1";
+		zephyr,reference = "ADC_REF_INTERNAL";
+		zephyr,acquisition-time = <ADC_ACQ_TIME_MAX>;
+		zephyr,resolution = <12>;
+	};
+};
+
+&i2c1 {
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/boards/nucleo_wb55rg.overlay:36]: If you're not enabling I2C and so on, this shouldn't be required

File: samples/boards/stm32/power_mgmt/adc/boards/nucleo_g474re.overlay [hunk 4]
```diff
@@ -0,0 +1,27 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * Copyright (c) 2020 Libre Solar Technologies GmbH
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/boards/nucleo_g474re.overlay:4]: Please update copyright.

File: samples/boards/stm32/power_mgmt/adc/README.rst [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+Build and flash Blinky as follows, changing ``nucleo_wb55rg`` for your board:
+
+.. zephyr-app-commands::
+   :zephyr-app: samples/boards/stm32/power_mgmt/adc
+   :board: nucleo_wb55rg
+   :goals: build flash
+   :compact:
+
+After flashing, the console shows the ADC measurements.
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/README.rst:32]: Provide a more detailed description

File: samples/boards/stm32/power_mgmt/adc/README.rst [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+PM configurations
+*****************
+
+By default, :kconfig:option:`CONFIG_PM_DEVICE` and :kconfig:option:`CONFIG_PM_DEVICE_RUNTIME` are
+enabled, but user can also deactivate one or the other to see each configuration
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/README.rst:38]: You' can't disable CONFIG_PM_DEVICE w/o disabling CONFIG_PM_DEVICE_RUNTIME

File: samples/boards/stm32/power_mgmt/adc/README.rst [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+in play.
```
Comment (erwango) [inline samples/boards/stm32/power_mgmt/adc/README.rst:39]: Maybe an example of the saving you're observing would be nice.

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/boards/stm32/power_mgmt/adc/prj.conf (hunk 2 above)
Comment (gautierg-st) [reply samples/boards/stm32/power_mgmt/adc/prj.conf:4]: I didn't see any changes in power consumption with exclusive enabled

**Review by erwango — CHANGES_REQUESTED** [review]

File: drivers/adc/adc_stm32.c [hunk 8]
```diff
@@ -1285,6 +1286,8 @@ static int adc_stm32_init(const struct device *dev)
 	ADC_TypeDef *adc = (ADC_TypeDef *)config->base;
 	int err;
 
+	ARG_UNUSED(adc); /* Necessary to avoid warnings on some series */
+
```
[the remark this answers is not in the document]
Comment (gautierg-st) [reply drivers/adc/adc_stm32.c:1290]: Done

**Review by erwango — APPROVED** [review]

**Review by FRASTM — APPROVED** [review]

**Pull request merged**